File: apps/light-desktop/src/features/programmingUpdate/contracts.ts
import type {
	UpdateMenuEntry,
	UpdateMode,
	UpdatePreview,
	UpdatePreviewItem,
	UpdateResult,
	UpdateSettings,
	UpdateTargetFilter,
	UpdateTargetIdentity,
	UpdateTargetRequest,
} from "../../api/types";
import type { ShowObject, ShowObjectBodies } from "../showObjects/contracts";

export type ProgrammingUpdateObjectKind = "cue_list" | "preset" | "group";

export type ProgrammingUpdateTarget =
	| {
			type: "cue";
			cue_list_id: string;
			playback_number?: number | null;
			cue_id?: string | null;
			cue_number?: number | null;
			validate_active_context: boolean;
	  }
	| { type: "preset"; object_id: string }
	| { type: "group"; object_id: string };

export interface ProgrammingUpdateObjectIdentity {
	kind: ProgrammingUpdateObjectKind;
	object_id: string;
	object_revision: number;
}

export interface ProgrammingUpdatePreview {
	target: UpdateTargetIdentity;
	mode: UpdateMode;
	items: UpdatePreviewItem[];
}

export interface ProgrammingUpdatePreviewRequest {
	request_id: string;
	target: ProgrammingUpdateTarget;
	mode: UpdateMode;
}

export interface ProgrammingUpdatePreviewResponse {
	request_id: string;
	correlation_id: string;
	show_id: string;
	show_revision: number;
	object: ProgrammingUpdateObjectIdentity;
	programmer_revision: string;
	preview: ProgrammingUpdatePreview;
}

export interface ProgrammingUpdateTargetsRequest {
	request_id: string;
	filter: UpdateTargetFilter;
}

export interface ProgrammingUpdateTargetEntry {
	request_target: ProgrammingUpdateTarget;
	object: ProgrammingUpdateObjectIdentity;
	programmer_revision: string;
	active_or_referenced: boolean;
	existing_preview: ProgrammingUpdatePreview;
	add_new_preview: ProgrammingUpdatePreview;
}

export interface ProgrammingUpdateTargetsResponse {
	request_id: string;
	correlation_id: string;
	show_id: string;
	show_revision: number;
	targets: ProgrammingUpdateTargetEntry[];
}

export type ProgrammingUpdateAction =
	| {
			type: "confirm_preview";
			target: ProgrammingUpdateTarget;
			mode: UpdateMode;
			expected_object_revision: number;
			expected_programmer_revision: string;
	  }
	| {
			type: "apply_direct";
			target: ProgrammingUpdateTarget;
			mode: UpdateMode;
	  };

export interface ProgrammingUpdateActionRequest {
	request_id: string;
	action: ProgrammingUpdateAction;
}

export type ProgrammingUpdateProjection = {
	[K in ProgrammingUpdateObjectKind]: {
		kind: K;
		object_id: string;
		object_revision: number;
		body: ShowObjectBodies[K];
	};
}[ProgrammingUpdateObjectKind];

export interface ProgrammingUpdateActionOutcome {
	status: "changed";
	request_id: string;
	correlation_id: string;
	replayed: boolean;
	show_id: string;
	show_revision: number;
	projection: ProgrammingUpdateProjection;
	event_sequence: number;
	summary: UpdateResult;
}

export interface ProgrammingUpdateSettingsProjection {
	desk_id: string;
	settings: UpdateSettings;
}

export interface ProgrammingUpdateTransport {
	preview(
		showId: string,
		request: ProgrammingUpdatePreviewRequest,
	): Promise<ProgrammingUpdatePreviewResponse>;
	targets(
		showId: string,
		request: ProgrammingUpdateTargetsRequest,
	): Promise<ProgrammingUpdateTargetsResponse>;
	apply(
		showId: string,
		expectedShowRevision: number,
		request: ProgrammingUpdateActionRequest,
	): Promise<ProgrammingUpdateActionOutcome>;
	loadSettings(deskId: string): Promise<ProgrammingUpdateSettingsProjection>;
	saveSettings(
		deskId: string,
		settings: UpdateSettings,
	): Promise<ProgrammingUpdateSettingsProjection>;
}

export class ProgrammingUpdateTransportError extends Error {
	constructor(
		message: string,
		readonly status: number,
		readonly currentShowRevision: number | null,
		readonly retryable: boolean,
	) {
		super(message);
		this.name = "ProgrammingUpdateTransportError";
	}
}

export interface ProgrammingUpdateScope {
	showId: string;
	deskId: string;
	userId: string;
	initialShowRevision: number | null;
}

export interface UpdatePreviewAuthority {
	scopeKey: string;
	requestId: string;
	correlationId: string;
	showId: string;
	showRevision: number;
	requestTarget: ProgrammingUpdateTarget;
	object: ProgrammingUpdateObjectIdentity;
	programmerRevision: string;
	preview: UpdatePreview;
}

export interface ProgrammingUpdateMenuEntry extends UpdateMenuEntry {
	requestTarget: ProgrammingUpdateTarget;
	object: ProgrammingUpdateObjectIdentity;
	programmerRevision: string;
	showRevision: number;
	existingAuthority: UpdatePreviewAuthority;
	addNewAuthority: UpdatePreviewAuthority;
}

export interface UpdateTargetsAuthority {
	scopeKey: string;
	requestId: string;
	correlationId: string;
	showId: string;
	showRevision: number;
	entries: ProgrammingUpdateMenuEntry[];
}

export type ProgrammingUpdateObject = {
	[K in ProgrammingUpdateObjectKind]: ShowObject<K> & { kind: K };
}[ProgrammingUpdateObjectKind];

export interface ProgrammingUpdateMutationResult {
	requestId: string;
	correlationId: string;
	replayed: boolean;
	showRevision: number;
	eventSequence: number;
	object: ProgrammingUpdateObject;
	result: UpdateResult;
}

export interface ProgrammingUpdateCapability {
	readonly scopeKey: string;
	loadSettings(): Promise<UpdateSettings | null>;
	saveSettings(settings: UpdateSettings): Promise<UpdateSettings | null>;
	preview(
		target: UpdateTargetRequest,
		mode: UpdateMode,
	): Promise<UpdatePreviewAuthority | null>;
	targets(filter: UpdateTargetFilter): Promise<UpdateTargetsAuthority | null>;
	confirm(
		authority: UpdatePreviewAuthority,
	): Promise<ProgrammingUpdateMutationResult | null>;
	applyDirect(
		target: UpdateTargetRequest,
		mode: UpdateMode,
	): Promise<ProgrammingUpdateMutationResult | null>;
}

File: crates/light/src/programming/update/contracts.rs
use super::{ActiveCueContext, UpdateMode, UpdatePreview, UpdateResult, UpdateTargetFilter};
use crate::{
    ActionContext, ActionError, ActiveShowObjectKind, ActiveShowPorts, ApplicationCommand,
    CommandFamily,
};
use light_core::{CueListId, Revision, ShowId};
use light_show::PortableShowRevision;
use std::sync::Arc;
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq)]
pub enum ProgrammingUpdateTargetRequest {
    Cue {
        cue_list_id: CueListId,
        playback_number: Option<u16>,
        cue_id: Option<Uuid>,
        cue_number: Option<f64>,
        validate_active_context: bool,
    },
    Preset {
        object_id: String,
    },
    Group {
        object_id: String,
    },
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdatePreviewRequest {
    pub show_id: ShowId,
    pub target: ProgrammingUpdateTargetRequest,
    pub mode: UpdateMode,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateTargetsRequest {
    pub show_id: ShowId,
    pub filter: UpdateTargetFilter,
}

impl ApplicationCommand for ProgrammingUpdateTargetsRequest {
    type Value = ProgrammingUpdateTargetsResult;

    const FAMILY: CommandFamily = CommandFamily::Programmer;
}

impl ApplicationCommand for ProgrammingUpdatePreviewRequest {
    type Value = ProgrammingUpdatePreviewResult;

    const FAMILY: CommandFamily = CommandFamily::Programmer;
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateCommand {
    pub show_id: ShowId,
    pub target: ProgrammingUpdateTargetRequest,
    pub mode: UpdateMode,
    pub expected_object_revision: Option<Revision>,
    pub expected_programmer_revision: Option<String>,
    pub expected_show_revision: Option<PortableShowRevision>,
}

impl ApplicationCommand for ProgrammingUpdateCommand {
    type Value = ProgrammingUpdateResult;

    const FAMILY: CommandFamily = CommandFamily::Programmer;
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdatePreviewResult {
    pub context: ActionContext,
    pub request_id: String,
    pub correlation_id: Uuid,
    pub show_revision: PortableShowRevision,
    pub object_revision: Revision,
    pub object: ProgrammingUpdateObjectReference,
    pub programmer_revision: String,
    pub preview: UpdatePreview,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateObjectReference {
    pub kind: ActiveShowObjectKind,
    pub object_id: String,
    pub object_revision: Revision,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateMenuEntry {
    pub target: ProgrammingUpdateTargetRequest,
    pub object_revision: Revision,
    pub object: ProgrammingUpdateObjectReference,
    pub programmer_revision: String,
    pub active_or_referenced: bool,
    pub existing_preview: UpdatePreview,
    pub add_new_preview: UpdatePreview,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateTargetsResult {
    pub context: ActionContext,
    pub request_id: String,
    pub correlation_id: Uuid,
    pub show_revision: PortableShowRevision,
    pub entries: Vec<ProgrammingUpdateMenuEntry>,
}

pub(crate) struct ProgrammingUpdateMenuInput {
    pub(crate) values: light_programmer::ProgrammerUpdateContent,
    pub(crate) selection: light_programmer::ProgrammerUpdateContent,
    pub(crate) values_fingerprint: String,
    pub(crate) selection_fingerprint: String,
    pub(crate) active_preset_id: Option<String>,
    pub(crate) referenced_group_ids: Vec<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateProjection {
    pub show_id: ShowId,
    pub kind: ActiveShowObjectKind,
    pub object_id: String,
    pub object_revision: Revision,
    pub raw_body: Arc<serde_json::Value>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateOutcome {
    pub projection: Arc<ProgrammingUpdateProjection>,
    pub show_revision: PortableShowRevision,
    pub event_sequence: u64,
    pub summary: UpdateResult,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProgrammingUpdateResult {
    pub context: ActionContext,
    pub request_id: String,
    pub correlation_id: Uuid,
    pub replayed: bool,
    pub outcome: ProgrammingUpdateOutcome,
}

pub trait ProgrammingUpdatePorts: ActiveShowPorts {
    fn authorize_programming_update(&self, context: &ActionContext) -> Result<(), ActionError>;

    fn active_update_cue_contexts(
        &self,
        context: &ActionContext,
    ) -> Result<Vec<ActiveCueContext>, ActionError>;

    fn reconcile_programming_update(&self, _projection: &ProgrammingUpdateProjection) {}
}
